File: android/src/main/kotlin/com/write4me/llama_flutter_android/LlamaFlutterAndroidPlugin.kt
package com.write4me.llama_flutter_android

import android.                )

                currentModelPath = config.modelPath
                isModelLoaded.set(true)
                withContext(Dispatchers.Main) {
                    callback(Result.success(Unit))
                }t.Context
import android.content.Intent
import androidx.core.content.ContextCompat
import io.flutter.embedding.engine.plugins.FlutterPlugin
import kotlinx.coroutines.*
import java.util.concurrent.atomic.AtomicBoolean

class LlamaFlutterAndroidPlugin : FlutterPlugin, LlamaHostApi {
    private lateinit var context: Context
    private lateinit var flutterApi: LlamaFlutterApi
    private val scope = CoroutineScope(Dispatchers.Default + SupervisorJob())
    private var generationJob: Job? = null
    private val isModelLoaded = AtomicBoolean(false)
    private val isStopping = AtomicBoolean(false)
    private var currentModelPath: String? = null

    companion object {
        init {
            System.loadLibrary("llama_jni")
        }
    }

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        flutterApi = LlamaFlutterApi(binding.binaryMessenger)
        LlamaHostApi.setUp(binding.binaryMessenger, this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        scope.cancel()
        if (isModelLoaded.get()) {
            nativeFreeModel()
        }
        LlamaHostApi.setUp(binding.binaryMessenger, null)
    }

    override fun loadModel(config: ModelConfig, callback: (Result<Unit>) -> Unit) {
        scope.launch {
            try {
                // Start foreground service for long-running task
                val intent = Intent(context, InferenceService::class.java)
                ContextCompat.startForegroundService(context, intent)

                // Load model with progress callback
                nativeLoadModel(
                    config.modelPath,
                    config.nThreads,
                    config.contextSize,
                    config.nGpuLayers ?: 0L
                ) { progress ->
                    scope.launch {
                        withContext(Dispatchers.Main) {
                            flutterApi.onLoadProgress(progress) { result ->
                                // Handle result if needed
                            }
                        }
                    }
                }

                isModelLoaded.set(true)
                withContext(Dispatchers.Main) {
                    callback(Result.success(Unit))
                }
            } catch (e: Exception) {
                scope.launch {
                    withContext(Dispatchers.Main) {
                        flutterApi.onError(e.message ?: "Failed to load model") { result ->
                            // Handle result if needed
                        }
                        callback(Result.failure(e))
                    }
                }
            }
        }
    }

    override fun generate(request: GenerateRequest, callback: (Result<Unit>) -> Unit) {
        if (!isModelLoaded.get()) {
            callback(Result.failure(IllegalStateException("Model not loaded")))
            return
        }

        isStopping.set(false)
        generationJob = scope.launch {
            try {
                nativeGenerate(
                    request.prompt,
                    request.maxTokens,
                    request.temperature,
                    request.topP,
                    request.topK
                ) { token ->
                    if (!isStopping.get()) {
                        scope.launch {
                            withContext(Dispatchers.Main) {
                                flutterApi.onToken(token) { result ->
                                    // Handle result if needed
                                }
                            }
                        }
                    }
                }

                if (!isStopping.get()) {
                    scope.launch {
                        withContext(Dispatchers.Main) {
                            flutterApi.onDone { result ->
                                // Handle result if needed
                            }
                        }
                    }
                }

                withContext(Dispatchers.Main) {
                    callback(Result.success(Unit))
                }
            } catch (e: Exception) {
                if (!isStopping.get()) {
                    scope.launch {
                        withContext(Dispatchers.Main) {
                            flutterApi.onError(e.message ?: "Generation failed") { result ->
                                // Handle result if needed
                            }
                            callback(Result.failure(e))
                        }
                    }
                }
            }
        }
    }

    override fun stop(callback: (Result<Unit>) -> Unit) {
        isStopping.set(true)
        generationJob?.cancel()
        nativeStop()
        callback(Result.success(Unit))
    }

    override fun dispose(callback: (Result<Unit>) -> Unit) {
        scope.launch {
            try {
                stop { }
                if (isModelLoaded.get()) {
                    nativeFreeModel()
                    isModelLoaded.set(false)
                }
                
                // Stop foreground service
                val intent = Intent(context, InferenceService::class.java)
                context.stopService(intent)
                
                withContext(Dispatchers.Main) {
                    callback(Result.success(Unit))
                }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) {
                    callback(Result.failure(e))
                }
            }
        }
    }

    override fun isModelLoaded(): Boolean {
        return isModelLoaded.get()
    }

    override fun generateChat(request: ChatRequest, callback: (Result<Unit>) -> Unit) {
        if (!isModelLoaded.get()) {
            callback(Result.failure(IllegalStateException("Model not loaded")))
            return
        }

        // Convert ChatMessage list to Kotlin ChatMessage list
        val messages = request.messages.map { msg ->
            com.write4me.llama_flutter_android.ChatMessage(
                role = msg.role,
                content = msg.content
            )
        }

        // Format using chat template
        val formattedPrompt = ChatTemplateManager.formatMessages(
            messages = messages,
            templateName = request.template,
            modelPath = currentModelPath
        )

        // Use existing generate method with formatted prompt
        val generateRequest = GenerateRequest(
            prompt = formattedPrompt,
            maxTokens = request.maxTokens,
            temperature = request.temperature,
            topP = request.topP,
            topK = request.topK
        )

        generate(generateRequest, callback)
    }

    override fun getSupportedTemplates(): List<String> {
        return ChatTemplateManager.getSupportedTemplates()
    }

    // Native methods
    private external fun nativeLoadModel(
        path: String,
        nThreads: Long,
        contextSize: Long,
        nGpuLayers: Long,
        progressCallback: (Double) -> Unit
    )

    private external fun nativeGenerate(
        prompt: String,
        maxTokens: Long,
        temperature: Double,
        topP: Double,
        topK: Long,
        tokenCallback: (String) -> Unit
    )

    private external fun nativeStop()
    private external fun nativeFreeModel()
}

File: ios/Classes/LlamaFlutterAndroidPlugin.swift
import Flutter
import Foundation

public class LlamaFlutterAndroidPlugin: NSObject, FlutterPlugin, LlamaHostApiProtocol {

    private var flutterApi: LlamaFlutterApi?
    private let wrapper = LlamaIosWrapper()
    private var isModelLoaded_ = false
    private var isStopping = false
    private var currentModelPath: String?
    private let queue = DispatchQueue(
        label: "com.write4me.llama_flutter_android.inference",
        qos: .userInitiated
    )

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = LlamaFlutterAndroidPlugin()
        instance.flutterApi = LlamaFlutterApi(binaryMessenger: registrar.messenger())
        LlamaHostApiSetup.setUp(binaryMessenger: registrar.messenger(), api: instance)
    }

    // MARK: - LlamaHostApiProtocol

    public func loadModel(config: ModelConfig, completion: @escaping (Result<Void, Error>) -> Void) {
        queue.async { [weak self] in
            guard let self = self else { return }
            do {
                try ObjcExceptionBridge.catch {
                    self.wrapper.loadModel(
                        atPath: config.modelPath,
                        nThreads: Int32(config.nThreads),
                        contextSize: Int32(config.contextSize),
                        nGpuLayers: Int32(config.nGpuLayers ?? 99),
                        progressCallback: { [weak self] progress in
                            DispatchQueue.main.async {
                                self?.flutterApi?.onLoadProgress(progress) {}
                            }
                        }
                    )
                }
                self.isModelLoaded_ = true
                self.currentModelPath = config.modelPath
                DispatchQueue.main.async { completion(.success(())) }
            } catch {
                DispatchQueue.main.async {
                    self.flutterApi?.onError(error.localizedDescription) {}
                    completion(.failure(error))
                }
            }
        }
    }

    public func generate(request: GenerateRequest, completion: @escaping (Result<Void, Error>) -> Void) {
        guard isModelLoaded_ else {
            completion(.failure(NSError(
                domain: "LlamaPlugin", code: 1,
                userInfo: [NSLocalizedDescriptionKey: "Model not loaded"]
            )))
            return
        }
        isStopping = false
        queue.async { [weak self] in
            guard let self = self else { return }
            do {
                try ObjcExceptionBridge.catch {
                    self.wrapper.generate(
                        withPrompt: request.prompt,
                        maxTokens: Int32(request.maxTokens),
                        temperature: request.temperature,
                        topP: request.topP,
                        topK: Int32(request.topK),
                        minP: request.minP,
                        typicalP: request.typicalP,
                        repeatPenalty: request.repeatPenalty,
                        frequencyPenalty: request.frequencyPenalty,
                        presencePenalty: request.presencePenalty,
                        repeatLastN: Int32(request.repeatLastN),
                        mirostat: Int32(request.mirostat),
                        mirostatTau: request.mirostatTau,
                        mirostatEta: request.mirostatEta,
                        seed: request.seed ?? -1,
                        penalizeNewline: request.penalizeNewline
                    ) { [weak self] token in
                        guard let self = self, !self.isStopping else { return }
                        DispatchQueue.main.async {
                            self.flutterApi?.onToken(token) {}
                        }
                    }
                }
                if !self.isStopping {
                    DispatchQueue.main.async {
                        self.flutterApi?.onDone {}
                        completion(.success(()))
                    }
                }
            } catch {
                if !self.isStopping {
                    DispatchQueue.main.async {
                        self.flutterApi?.onError(error.localizedDescription) {}
                        completion(.failure(error))
                    }
                }
            }
        }
    }

    public func generateChat(request: ChatRequest, completion: @escaping (Result<Void, Error>) -> Void) {
        guard isModelLoaded_ else {
            completion(.failure(NSError(
                domain: "LlamaPlugin", code: 1,
                userInfo: [NSLocalizedDescriptionKey: "Model not loaded"]
            )))
            return
        }
        isStopping = false

        let messages = request.messages.map { TemplateChatMessage(role: $0.role, content: $0.content) }
        let formattedPrompt = ChatTemplateManager.shared.formatMessages(
            messages: messages,
            templateName: request.template,
            modelPath: currentModelPath
        )

        let generateRequest = GenerateRequest(
            prompt: formattedPrompt,
            maxTokens: request.maxTokens,
            temperature: request.temperature,
            topP: request.topP,
            topK: request.topK,
            minP: request.minP,
            typicalP: request.typicalP,
            repeatPenalty: request.repeatPenalty,
            frequencyPenalty: request.frequencyPenalty,
            presencePenalty: request.presencePenalty,
            repeatLastN: request.repeatLastN,
            mirostat: request.mirostat,
            mirostatTau: request.mirostatTau,
            mirostatEta: request.mirostatEta,
            seed: request.seed,
            penalizeNewline: request.penalizeNewline
        )
        generate(request: generateRequest, completion: completion)
    }

    public func getSupportedTemplates() -> [String] {
        return ChatTemplateManager.shared.getSupportedTemplates()
    }

    public func stop(completion: @escaping (Result<Void, Error>) -> Void) {
        isStopping = true
        wrapper.stop()
        completion(.success(()))
    }

    public func dispose(completion: @escaping (Result<Void, Error>) -> Void) {
        isStopping = true
        wrapper.stop()
        queue.async { [weak self] in
            guard let self = self else { return }
            self.wrapper.freeModel()
            self.isModelLoaded_ = false
            DispatchQueue.main.async { completion(.success(())) }
        }
    }

    public func isModelLoaded() -> Bool {
        return isModelLoaded_
    }

    public func getContextInfo() -> ContextInfo {
        let used = Int64(wrapper.tokensUsed())
        let size = Int64(wrapper.contextSize())
        let pct  = size > 0 ? (Double(used) / Double(size) * 100.0) : 0.0
        return ContextInfo(tokensUsed: used, contextSize: size, usagePercentage: pct)
    }

    public func clearContext(completion: @escaping (Result<Void, Error>) -> Void) {
        queue.async { [weak self] in
            self?.wrapper.clearContext()
            DispatchQueue.main.async { completion(.success(())) }
        }
    }

    public func setSystemPromptLength(length: Int64) {
        wrapper.setSystemPromptLength(Int32(length))
    }

    public func registerCustomTemplate(name: String, content: String) {
        ChatTemplateManager.shared.registerCustomTemplate(name: name, content: content)
    }

    public func unregisterCustomTemplate(name: String) {
        ChatTemplateManager.shared.unregisterCustomTemplate(name: name)
    }
}
